Compiling NITSOL2

NITSOL is an iterative solver for large systems of nonlinear equations. A description can be found in:

> Pernice, M., & Walker, H. F. (1998). NITSOL: A Newton iterative solver for nonlinear systems. SIAM Journal on Scientific Computing, 19(1), 302-318. https://doi.org/10.1137/S1064827596303843

The original source code can be found on H. Walker's homepage: https://users.wpi.edu/~walker/NITSOL/

A modified version that removes COMMON blocks can be found on GitHub: https://github.com/vadim-z/NITSOL2

```
git clone https://github.com/vadim-z/NITSOL2 && cd NITSOL2
cmake -B build -DCMAKE_Fortran_COMPILER=`which lfortran` -DCMAKE_Fortran_FLAGS="--fixed-form --implicit-interface --implicit-typing --legacy-array-sections"
cmake --build build -v
```

Result compiling with LFortran version 0.62: 

```
...
gmake[2]: Entering directory '/Users/ivan/fortran/NITSOL2/build'
[  2%] Building Fortran object CMakeFiles/nitsol2.dir/src/nitdflts.f.o
/opt/homebrew/Cellar/micromamba/2.5.0_4/bin/lfortran  -I/Users/ivan/fortran/NITSOL2 --fixed-form --implicit-interface --implicit-typing --legacy-array-sections -c /Users/ivan/fortran/NITSOL2/src/nitdflts.f -o CMakeFiles/nitsol2.dir/src/nitdflts.f.o
[  5%] Building Fortran object CMakeFiles/nitsol2.dir/src/nitjv.f.o
/opt/homebrew/Cellar/micromamba/2.5.0_4/bin/lfortran  -I/Users/ivan/fortran/NITSOL2 --fixed-form --implicit-interface --implicit-typing --legacy-array-sections -c /Users/ivan/fortran/NITSOL2/src/nitjv.f -o CMakeFiles/nitsol2.dir/src/nitjv.f.o
[  7%] Building Fortran object CMakeFiles/nitsol2.dir/src/nitcgm.f.o
/opt/homebrew/Cellar/micromamba/2.5.0_4/bin/lfortran  -I/Users/ivan/fortran/NITSOL2 --fixed-form --implicit-interface --implicit-typing --legacy-array-sections -c /Users/ivan/fortran/NITSOL2/src/nitcgm.f -o CMakeFiles/nitsol2.dir/src/nitcgm.f.o
gmake[2]: *** [CMakeFiles/nitsol2.dir/build.make:101: CMakeFiles/nitsol2.dir/src/nitcgm.f.o] Segmentation fault: 11
gmake[2]: Leaving directory '/Users/ivan/fortran/NITSOL2/build'
gmake[1]: *** [CMakeFiles/Makefile2:136: CMakeFiles/nitsol2.dir/all] Error 2
gmake[1]: Leaving directory '/Users/ivan/fortran/NITSOL2/build'
gmake: *** [Makefile:91: all] Error 2
```
Shorter command to reproduce the error:

```
$ lfortran --fixed-form --implicit-interface --legacy-array-sections -c src/nitcgm.f
Segmentation fault: 11
```

Curiously, with version 0.59 no segfault occurs: https://godbolt.org/z/1q9z6sYco

--- a/integration_tests/CMakeLists.txt
+++ b/integration_tests/CMakeLists.txt
@@ -2010,6 +2010,7 @@ RUN(NAME implicit_interface_33 LABELS gfortran llvm EXTRA_ARGS --implicit-interf
 RUN(NAME implicit_interface_34 LABELS gfortran llvm EXTRA_ARGS --implicit-interface)
 RUN(NAME implicit_interface_35 LABELS gfortran llvm EXTRA_ARGS --implicit-typing --implicit-interface)
 RUN(NAME implicit_interface_36 LABELS gfortran llvm EXTRA_ARGS --implicit-interface)
+RUN(NAME implicit_interface_37 LABELS gfortran llvm EXTRA_ARGS --implicit-interface --legacy-array-sections)
 
 RUN(NAME implicit_typing_01 LABELS gfortran llvmImplicit)
 RUN(NAME implicit_typing_02 LABELS gfortran llvmImplicit)

--- a/integration_tests/implicit_interface_37.f90
+++ b/integration_tests/implicit_interface_37.f90
@@ -1,0 +1,33 @@
+subroutine compute(a, b, proc)
+integer :: a(2), b(1)
+external proc
+call sub1(b, proc)
+call sub1(a(2), proc)
+end subroutine
+
+subroutine sub1(x, proc)
+integer :: x(*)
+external proc
+call proc(x(1))
+end subroutine
+
+subroutine add_one(val)
+integer :: val
+val = val + 1
+end subroutine
+
+program implicit_interface_37
+implicit none
+integer :: a(2), b(1)
+external add_one
+
+a = [10, 20]
+b = [5]
+
+call compute(a, b, add_one)
+
+if (b(1) /= 6) error stop
+if (a(2) /= 21) error stop
+
+print *, "ok"
+end program

--- a/src/libasr/codegen/asr_to_llvm.cpp
+++ b/src/libasr/codegen/asr_to_llvm.cpp
@@ -23628,7 +23628,9 @@ public:
                 llvm::BasicBlock *optional_check_mergeBB = nullptr;
                 if (i + 1 < x.n_args && x.m_args[i + 1].m_value != nullptr) {
                     ASR::expr_t* next_arg = x.m_args[i + 1].m_value;
-                    if (ASR::is_a<ASR::Var_t>(*next_arg)) {
+                    if (ASR::is_a<ASR::Var_t>(*next_arg) &&
+                        ASR::is_a<ASR::Variable_t>(*ASRUtils::symbol_get_past_external(
+                            ASR::down_cast<ASR::Var_t>(next_arg)->m_v))) {
                         ASR::Variable_t* next_var = ASRUtils::EXPR2VAR(next_arg);
                         std::string var_name = std::string(next_var->m_name);
                         ASR::ttype_t* var_type = ASRUtils::type_get_past_pointer(next_var->m_type);

